igor-web/src/test/groovy/com/netflix/spinnaker/igor/wercker/WerckerBuildMonitorSpec.groovy

```diff
@@ -0,0 +1,251 @@
+ *
+ * The contents of this file are subject to the Apache License Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * If a copy of the Apache License Version 2.0 was not distributed with this file,
+ * You can obtain one at https://www.apache.org/licenses/LICENSE-2.0.html
+ */
+package com.netflix.spinnaker.igor.wercker
+
+import com.netflix.spectator.api.NoopRegistry
+import com.netflix.spinnaker.igor.IgorConfigurationProperties
+import com.netflix.spinnaker.igor.config.WerckerProperties
+import com.netflix.spinnaker.igor.config.WerckerProperties.WerckerHost
+import com.netflix.spinnaker.igor.history.EchoService
+import com.netflix.spinnaker.igor.model.BuildServiceProvider
+import com.netflix.spinnaker.igor.service.BuildMasters
+import com.netflix.spinnaker.igor.wercker.model.Application
+import com.netflix.spinnaker.igor.wercker.model.Owner
+import com.netflix.spinnaker.igor.wercker.model.Pipeline
+import com.netflix.spinnaker.igor.wercker.model.Run
+import com.netflix.spinnaker.kork.eureka.RemoteStatusChangedEvent
+
+import java.util.concurrent.TimeUnit
+
+import rx.schedulers.TestScheduler
+import spock.lang.Specification
+
+class WerckerBuildMonitorSpec extends Specification {
+
+    WerckerCache cache = Mock(WerckerCache)
+    WerckerBuildMonitor monitor
+    EchoService echoService = Mock(EchoService)
+    WerckerClient client
+    WerckerService mockService = Mock(WerckerService)
+    WerckerService werckerService
+    String werckerDev = 'https://dev.wercker.com/'
+    String master = 'WerckerTestMaster'
+
+    void setup() {
+        client = Mock(WerckerClient)
+        werckerService = new WerckerService(
+                new WerckerHost(name: master, address: werckerDev), cache, client)
+    }
+
+    final MASTER = 'MASTER'
+    final pipeline = 'myOrg/myApp/pollingTest'
+    final List<String> APPS = []
+    final TestScheduler scheduler = new TestScheduler()
+
+    void 'no finished run'() {
+        given:
+        cache.getJobNames(MASTER) >> ['pipeline']
+        BuildMasters buildMasters = Mock(BuildMasters)
+        buildMasters.map >> [MASTER: mockService]
+
+        long now = System.currentTimeMillis();
+        List<Run> runs1 = [
+            new Run(id:"b",    startedAt: new Date(now-10)),
+            new Run(id:"a",    startedAt: new Date(now-11)),
+            new Run(id:"init", startedAt: new Date(now-12)),
+        ]
+        monitor = monitor(buildMasters)
+        monitor.worker = scheduler.createWorker()
+        mockService.getRunsSince(_) >> [pipeline: runs1]
+        cache.getBuildNumber(_, _, _) >> 1
+
+        when:
+        monitor.onApplicationEvent(Mock(RemoteStatusChangedEvent))
+        scheduler.advanceTimeBy(1L, TimeUnit.SECONDS.MILLISECONDS)
+
+        then: 'initial poll'
+        1 * buildMasters.filteredMap(BuildServiceProvider.WERCKER) >> [MASTER: mockService]
+        1 * buildMasters.map >> [MASTER: mockService]
```

Comment:
This looks to be duplicating the mock in the `given` section.  If we're stubbing (and don't really care about testing the number of calls) we can move it to `given`; if testing the number of calls is part of the test then it should stay in `then`.  (My guess is that the number of calls to `buildMasters` is an implementation detail so it should go in `given`)

Reply:
Moved the duplicated setup to mockBuildMasters().